keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,1330 @@
+setting the `EvacuationInitiated` condition on the pod.
+
+The node maintenance controller would also remove the `EvacuationRequested` condition from the
+targeted pods if the `.spec.stage` of the NodeMaintenance is set to `Complete`, unless there is no
+other maintenance in progress. The condition will only be removed if the reason of the condition
+is `NodeMaintenance$NAME`, where NAME is the name of the NodeMaintenance. If the reason has a
+different value, then it is owned by another controller (e.g. descheduler) or by a different
+NodeMaintenance object and we should keep the condition.
+
+The controller can show progress by reconciling:
+- `.status.nodes[0].PodsPendingEvacuation`, to show how many pods remain to be removed from
+  the first node.
+- `.status.nodes[0].PodsEvacuating`, to show how many pods have been accepted for the
+  evacuation from the first node. These are the pods that have the `EvacuationInitiated`
+  condition set to `True`.
+- To keep track of the entire maintenance the controller will reconcile a `Drained` condition and
+  set it to true if all pods pending evacuation/termination have terminated on all target nodes
+  when drain is requested by the maintenance object.
+- NodeMaintenance condition or annotation can be set on the node object to advertise the current
+  phase of the maintenance.
+
+#### Supported Stage Transitions
+
+The following transitions should be validated by the API server.
+
+- Idle -> _Deletion_
+  - Planning a maintenance in the future and canceling/deleting it without any consequence.
+- (Idle) -> Cordon -> (Complete) -> _Deletion_.
+  - Make a set of nodes unschedulable and then schedulable again.
+  - The complete stage will always be run even without specifying it.
+- (Idle) -> (Cordon) -> Drain -> (Complete) -> _Deletion_.
+  - Make a set of nodes unschedulable, drain them, and then make them schedulable again.
+  - Cordon and Complete stages will always be run, even without specifying them.
+- (Idle) -> Complete -> _Deletion_.
+  -  Make a set of nodes schedulable.
+
+### Deployment and ReplicaSet Controllers
+
+The replica set controller will watch its pods and count the number of pods it observes with a
+`EvacuationRequested` condition. It will then store this count in `.status.ReplicasToEvacuate`
+
+The deployment controller will watch its ReplicaSets and react when it observes positive number of
+pods in `.status.ReplicasToEvacuate`. If the owning object of the targeted pods is a Deployment with
+a positive `.spec.strategy.rollingUpdate.maxSurge` value, the controller will create surge pods by
+scaling up the ReplicaSet. The new pods will not be scheduled on the maintained node because the
+`.spec.unschedulable` field would be set to true on that node. As soon as the surge pods become
+available, the deployment controller will scale down the ReplicaSet. The replica set controller
+will then in turn delete the pods with the `EvacuationRequested` condition.
```

Comment:
Could you please also spell out what happens if `maxSurge` is `0` and when the deployment uses the `Recreate` strategy?

Reply:
The Deployment controller has been moved to the Evacuation API KEP https://github.com/kubernetes/enhancements/pull/4565. PTAL